cpp/src/experimental/random_walks.cuh

```diff
@@ -0,0 +1,854 @@
+                     device_vec_t<index_t> const& d_sizes,
+                     index_t num_paths,
+                     index_t max_depth)
+    : handle_(handle),
+      col_indices_(raw_const_ptr(d_indices)),
+      row_offsets_(raw_const_ptr(d_offsets)),
+      values_(raw_const_ptr(d_values)),
+      out_degs_(raw_const_ptr(d_crt_out_degs)),
+      sizes_(raw_const_ptr(d_sizes)),
+      num_paths_(num_paths),
+      max_depth_(max_depth)
+  {
+  }
+
+  col_indx_extract_t(raft::handle_t const& handle,
+                     vertex_t const* p_d_indices,
+                     edge_t const* p_d_offsets,
+                     weight_t const* p_d_values,
+                     edge_t const* p_d_crt_out_degs,
+                     index_t const* p_d_sizes,
+                     index_t num_paths,
+                     index_t max_depth)
+    : handle_(handle),
+      col_indices_(p_d_indices),
+      row_offsets_(p_d_offsets),
+      values_(p_d_values),
+      out_degs_(p_d_crt_out_degs),
+      sizes_(p_d_sizes),
+      num_paths_(num_paths),
+      max_depth_(max_depth)
+  {
+  }
+
+  col_indx_extract_t(raft::handle_t const& handle,
+                     graph_t const& graph,
+                     edge_t const* p_d_crt_out_degs,
+                     index_t const* p_d_sizes,
+                     index_t num_paths,
+                     index_t max_depth)
+    : handle_(handle),
+      col_indices_(graph.indices()),
+      row_offsets_(graph.offsets()),
+      values_(graph.weights()),
+      out_degs_(p_d_crt_out_degs),
+      sizes_(p_d_sizes),
+      num_paths_(num_paths),
+      max_depth_(max_depth)
+  {
+  }
+
+  // in-place extractor of next set of vertices and weights,
+  // (d_v_next_vertices, d_v_next_weights),
+  // given start set of vertices. d_v_src_vertices,
+  // and corresponding column index set, d_v_col_indx:
+  //
+  // for each indx in [0, num_paths){
+  //   v_indx = d_v_src_vertices[indx*max_depth + d_sizes[indx] - 1];
+  //   if( out_degs_[v_indx] > 0 ) {
+  //      start_row = row_offsets_[v_indx];
+  //      delta = d_v_col_indx[indx];
+  //      d_v_next_vertices[indx] = col_indices_[start_row + delta];
+  // }
+  // (use tranform_if() with transform iterator)
+  //
+  void operator()(
+    device_vec_t<vertex_t> const& d_coalesced_src_v,  // in: coalesced vector of vertices
+    device_vec_t<vertex_t> const&
+      d_v_col_indx,  // in: column indices, given by stepper's random engine
+    device_vec_t<vertex_t>& d_v_next_vertices,  // out: set of destination vertices, for next step
+    device_vec_t<weight_t>&
+      d_v_next_weights)  // out: set of weights between src and destination vertices, for next step
+    const
+  {
+    using zip_iterator_t =
+      thrust::zip_iterator<thrust::tuple<device_v_it<vertex_t>, device_v_it<weight_t>>>;
+
+    // auto max_depth          = max_depth_;  // to avoid capturing `this`...
+    // auto const* ptr_d_sizes = sizes_;
+
+    thrust::transform_if(
```

Comment:
Does this work if there are 0 degree vertices, shouldn't we need to take the return value of transform_if to figure out how many column indices are actually extracted?

Reply:
Per another related comment, the bookkeeping of which column indices are extracted are maintained by `d_crt_out_degs`: wherever there are 0 entries, no transformation is done, and corresponding entries in target vector are not updated.

Reply:
And, yes, current tests show that, as they have sinks and/or vertices with out-degree less than max_depth and the stopping condition detects 0-degree vertices and acts accordingly.